### PR TITLE
BUG: transform with nunique should have dtype int64

### LINKED ISSUE #35109: BUG: Groupby.transform on datetime with nunique giving incorrect result
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
# Your code here
>>> import pandas as pd
>>> bdates = pd.bdate_range('2008-12-31', '2011-12-30')
>>> df = pd.DataFrame(bdates, columns=['date'])
>>> df['year'] = df['date'].dt.year
>>> df.groupby('year')['date'].transform('nunique')
0     1970-01-01 00:00:00.000000001
1     1970-01-01 00:00:00.000000261
2     1970-01-01 00:00:00.000000261
3     1970-01-01 00:00:00.000000261
4     1970-01-01 00:00:00.000000261
                   ...             
778   1970-01-01 00:00:00.000000260
779   1970-01-01 00:00:00.000000260
780   1970-01-01 00:00:00.000000260
781   1970-01-01 00:00:00.000000260
782   1970-01-01 00:00:00.000000260
Name: date, Length: 783, dtype: datetime64[ns]
>>> 

```

#### Problem description

The current behaviour returns the sum of default timestamp with the number of unique count of dates in each year. The expected output should have been the number of unique counts for the date in each year which is consistent with the behavior of other Groupby.transform functions.

#### Expected Output

```python
# Your code here
>>> import pandas as pd
>>> bdates = pd.bdate_range('2008-12-31', '2011-12-30')
>>> df = pd.DataFrame(bdates, columns=['date'])
>>> df['year'] = df['date'].dt.year
>>> df.groupby('year')['date'].transform('nunique')
0     1
1     261
2     261
3     261
4     261
                   ...             
778   260
779   260
780   260
781   260
782   260
Name: date, Length: 783, dtype: int
>>> 

```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.7.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.14.181-140.257.amzn2.x86_64
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.4
numpy            : 1.18.5
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 47.1.1.post20200604
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.15.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.2.1
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.4
pandas_gbq       : None
pyarrow          : 0.17.1
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : 0.8.7
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
</details>

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #35109
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -319,6 +319,9 @@ def maybe_cast_result_dtype(dtype: DtypeObj, how: str) -> DtypeObj:
         return np.dtype(np.int64)
     elif how in ["add", "cumsum", "sum"] and isinstance(dtype, BooleanDtype):
         return Int64Dtype()
+    elif how == "nunique":
```
Comment (WillAyd) [inline pandas/core/dtypes/cast.py:322]: Do you know where in nunique this is getting cast to date time? Not sure this generic function is the best place to handle this

Comment (rhshadrach) [reply pandas/core/dtypes/cast.py:322]: The actual casting is done on the last line of this block:

https://github.com/pandas-dev/pandas/blob/74538104bc04238f69ea6c3a6fbc5468fe02ed12/pandas/core/dtypes/cast.py#L156-L166

This function was added to allow overrides to the default casting behavior based on input datatype and the operation being performed, which I think is what we want to do here.

Comment (WillAyd) [reply pandas/core/dtypes/cast.py:322]: Is that being hit through `_transform_general`? If so seems like something else if off as this should be a reduction function

Comment (rhshadrach) [reply pandas/core/dtypes/cast.py:322]: Ah - I think I understand your comment better now: perhaps there is some case in this block that is casting but shouldn't be. In the case at hand, `dtype.kind` is `"M"` and `result.dtype.kind` is `"i"`. Here is an example from the tests where we want to execute the last line:

    df = pd.DataFrame(dict(a=[1], b=pd.to_datetime(["2001"])))
    df.groupby("a")["b"].cummin()

Comment (WillAyd) [reply pandas/core/dtypes/cast.py:322]: Right; ideally we don't have to special case here after we've already cast things. IIUC something seems off with the code path in the first place if we are going through transform machinery to get to this function, as nunique should be a reduction not a transformation

Comment (rhshadrach) [reply pandas/core/dtypes/cast.py:322]: The function `maybe_cast_result` is being called by `_transform_fast` for both frames and series:

https://github.com/pandas-dev/pandas/blob/74538104bc04238f69ea6c3a6fbc5468fe02ed12/pandas/core/groupby/generic.py#L1496-L1502

https://github.com/pandas-dev/pandas/blob/74538104bc04238f69ea6c3a6fbc5468fe02ed12/pandas/core/groupby/generic.py#L549-L552

It seems that the typical behavior is to have transform cast: `_transform_should_cast` returns true any time the result is nonempty and the function name is not in `base.cython_cast_blocklist`.

Comment (rhshadrach) [reply pandas/core/dtypes/cast.py:322]: Your point is well taken though - if we're taking the result of a reduction and broadcasting to the entire frame/series, why do we need to cast? Taking these casts out, the only test that fails is 

`groupby.transform.test_transform.test_categorical_and_not_categorical_key`.

I'm going to take a deeper look into this case and see what's going on; removing the casting would be a much better solution. Thanks for the comments.

**Comment by jreback** [comment]
superseded by #35152

**Pull request closed without merging**